File: DSCI553/hw4/task2.py
import sys
from pyspark import SparkContext
from collections import defaultdict
from copy import deepcopy

def compute_edge_betweenness(adjacency_map, node_list):
    edge_betweenness_scores = defaultdict(float)
    
    for source_node in node_list:
        predecessor_map = defaultdict(set)
        distance_map = {}
        shortest_path_count = defaultdict(float)
        traversal_order = []
        bfs_queue = []
        
        bfs_queue.append(source_node)
        explored_nodes = set()
        explored_nodes.add(source_node)
        distance_map[source_node] = 0
        shortest_path_count[source_node] = 1
        
        while len(bfs_queue) > 0:
            current_node = bfs_queue.pop(0)
            traversal_order.append(current_node)
            
            for adjacent_node in adjacency_map[current_node]:
                if adjacent_node not in explored_nodes:
                    bfs_queue.append(adjacent_node)
                    explored_nodes.add(adjacent_node)
                    
                    if adjacent_node not in predecessor_map.keys():
                        predecessor_map[adjacent_node] = set()
                        predecessor_map[adjacent_node].add(current_node)
                    else:
                        predecessor_map[adjacent_node].add(current_node)
                    
                    shortest_path_count[adjacent_node] += shortest_path_count[current_node]
                    distance_map[adjacent_node] = distance_map[current_node] + 1
                    
                elif distance_map[adjacent_node] == distance_map[current_node] + 1:
                    if adjacent_node not in predecessor_map.keys():
                        predecessor_map[adjacent_node] = set()
                        predecessor_map[adjacent_node].add(current_node)
                    else:
                        predecessor_map[adjacent_node].add(current_node)
                    
                    shortest_path_count[adjacent_node] += shortest_path_count[current_node]
        
        node_weight_map = {}
        for node in traversal_order:
            node_weight_map[node] = 1
        
        edge_contribution = defaultdict(float)
        
        for node in reversed(traversal_order):
            for predecessor in predecessor_map[node]:
                contribution = node_weight_map[node] * (shortest_path_count[predecessor] / shortest_path_count[node])
                node_weight_map[predecessor] += contribution
                
                edge_key = tuple(sorted([node, predecessor]))
                if edge_key not in edge_contribution.keys():
                    edge_contribution[edge_key] = contribution
                else:
                    edge_contribution[edge_key] += contribution
        
        for edge, contribution_value in edge_contribution.items():
            if edge not in edge_betweenness_scores.keys():
                edge_betweenness_scores[edge] = contribution_value / 2
            else:
                edge_betweenness_scores[edge] += contribution_value / 2
    
    sorted_betweenness = sorted(edge_betweenness_scores.items(), key=lambda x: (-x[1], x[0]))
    return sorted_betweenness


if __name__ == '__main__':
    threshold_value = sys.argv[1]
    data_input_path = sys.argv[2]
    betweenness_file_path = sys.argv[3]
    community_file_path = sys.argv[4]
    
    spark_context = SparkContext('local[*]', 'task2')
    spark_context.setLogLevel('ERROR')
    
    data_lines = spark_context.textFile(data_input_path)
    header_line = data_lines.first()
    data_lines = data_lines.filter(lambda row: row != header_line).map(lambda row: row.split(","))
    
    user_to_businesses = data_lines.groupByKey().mapValues(set)
    business_set_by_user = {}
    for uid, bids in user_to_businesses.collect():
        business_set_by_user[uid] = bids
    
    all_users = data_lines.map(lambda row: row[0]).distinct()
    user_pairs = []
    user_collection = all_users.collect()
    
    for i in range(len(user_collection)):
        for j in range(i + 1, len(user_collection)):
            user_pairs.append((user_collection[i], user_collection[j]))
            user_pairs.append((user_collection[j], user_collection[i]))
    
    edge_list = []
    vertex_collection = set()
    for user_a, user_b in user_pairs:
        common_businesses = business_set_by_user[user_a] & business_set_by_user[user_b]
        if len(common_businesses) >= int(threshold_value):
            edge_list.append((user_a, user_b))
            vertex_collection.add(user_a)
    
    vertices = list(vertex_collection)
    
    adjacency_structure = {}
    for user_x, user_y in edge_list:
        if user_x not in adjacency_structure.keys():
            adjacency_structure[user_x] = set()
            adjacency_structure[user_x].add(user_y)
        else:
            adjacency_structure[user_x].add(user_y)
        
        if user_y not in adjacency_structure.keys():
            adjacency_structure[user_y] = set()
            adjacency_structure[user_y].add(user_x)
        else:
            adjacency_structure[user_y].add(user_x)
    
    betweenness_results = compute_edge_betweenness(adjacency_structure, vertices)
    
    with open(betweenness_file_path, "w") as output_file:
        for edge_tuple, score in betweenness_results:
            edge_str = str(edge_tuple)
            output_line = edge_str + "," + str(round(score, 5)) + "\n"
            output_file.write(output_line)
    
    working_graph = deepcopy(adjacency_structure)
    total_edges = len(betweenness_results)
    
    node_degrees = {node: len(adjacency_structure[node]) for node in adjacency_structure}
    
    best_modularity_score = -float('inf')
    optimal_communities = []
    
    while len(betweenness_results) > 0:
        component_list = []
        remaining_vertices = vertices.copy()
        
        while len(remaining_vertices) > 0:
            start_vertex = remaining_vertices.pop()
            component_queue = []
            component_queue.append(start_vertex)
            visited_set = set()
            visited_set.add(start_vertex)
            
            while len(component_queue) > 0:
                vertex = component_queue.pop(0)
                for neighbor in working_graph[vertex]:
                    if neighbor not in visited_set:
                        remaining_vertices.remove(neighbor)
                        component_queue.append(neighbor)
                        visited_set.add(neighbor)
            
            component_members = sorted(list(visited_set))
            component_list.append(component_members)
        
        modularity_value = 0.0
        for component in component_list:
            for node_i in component:
                for node_j in component:
                    has_edge = 1.0 if node_j in adjacency_structure[node_i] else 0.0
                    modularity_value += has_edge - (node_degrees[node_i] * node_degrees[node_j]) / (2.0 * total_edges)
        
        modularity_value /= (2 * total_edges)
        
        if modularity_value > best_modularity_score:
            best_modularity_score = modularity_value
            optimal_communities = deepcopy(component_list)
        
        max_betweenness = betweenness_results[0][1]
        for edge, betweenness_val in betweenness_results:
            if betweenness_val >= max_betweenness:
                working_graph[edge[0]].remove(edge[1])
                working_graph[edge[1]].remove(edge[0])
            
        betweenness_results = compute_edge_betweenness(working_graph, vertices)
    
    optimal_communities = sorted(optimal_communities, key=lambda x: (len(x), x[0]))
    
    with open(community_file_path, "w") as output_file:
        for community in optimal_communities:
            community_str = str(community)
            output_line = community_str[1:-1] + "\n"
            output_file.write(output_line)
    
    spark_context.stop()

File: DSCI553/hw4/task2.scala
import org.apache.spark.SparkContext
import org.apache.spark.rdd.RDD
import scala.collection.mutable
import scala.collection.mutable.{Map => MutableMap, Set => MutableSet, ArrayBuffer, Queue}

object task2 {
  
  def computeEdgeBetweenness(
    adjacencyMap: Map[String, Set[String]], 
    nodeList: List[String]
  ): List[((String, String), Double)] = {
    
    val edgeBetweennessScores = MutableMap[(String, String), Double]()
    
    for (sourceNode <- nodeList) {
      val predecessorMap = MutableMap[String, MutableSet[String]]()
      val distanceMap = MutableMap[String, Int]()
      val shortestPathCount = MutableMap[String, Double]()
      val traversalOrder = ArrayBuffer[String]()
      val bfsQueue = Queue[String]()
      
      bfsQueue.enqueue(sourceNode)
      val exploredNodes = MutableSet[String]()
      exploredNodes.add(sourceNode)
      distanceMap(sourceNode) = 0
      shortestPathCount(sourceNode) = 1.0
      
      while (bfsQueue.nonEmpty) {
        val currentNode = bfsQueue.dequeue()
        traversalOrder += currentNode
        
        for (adjacentNode <- adjacencyMap(currentNode)) {
          if (!exploredNodes.contains(adjacentNode)) {
            bfsQueue.enqueue(adjacentNode)
            exploredNodes.add(adjacentNode)
            
            if (!predecessorMap.contains(adjacentNode)) {
              predecessorMap(adjacentNode) = MutableSet[String]()
            }
            predecessorMap(adjacentNode).add(currentNode)
            
            shortestPathCount(adjacentNode) = shortestPathCount.getOrElse(adjacentNode, 0.0) + 
                                              shortestPathCount(currentNode)
            distanceMap(adjacentNode) = distanceMap(currentNode) + 1
            
          } else if (distanceMap(adjacentNode) == distanceMap(currentNode) + 1) {
            if (!predecessorMap.contains(adjacentNode)) {
              predecessorMap(adjacentNode) = MutableSet[String]()
            }
            predecessorMap(adjacentNode).add(currentNode)
            
            shortestPathCount(adjacentNode) = shortestPathCount(adjacentNode) + 
                                              shortestPathCount(currentNode)
          }
        }
      }
      
      val nodeWeightMap = MutableMap[String, Double]()
      for (node <- traversalOrder) {
        nodeWeightMap(node) = 1.0
      }
      
      val edgeContribution = MutableMap[(String, String), Double]()
      
      for (node <- traversalOrder.reverse) {
        if (predecessorMap.contains(node)) {
          for (predecessor <- predecessorMap(node)) {
            val contribution = nodeWeightMap(node) * 
                             (shortestPathCount(predecessor) / shortestPathCount(node))
            nodeWeightMap(predecessor) = nodeWeightMap(predecessor) + contribution
            
            val edgeKey = if (node < predecessor) (node, predecessor) else (predecessor, node)
            edgeContribution(edgeKey) = edgeContribution.getOrElse(edgeKey, 0.0) + contribution
          }
        }
      }
      
      for ((edge, contributionValue) <- edgeContribution) {
        edgeBetweennessScores(edge) = edgeBetweennessScores.getOrElse(edge, 0.0) + 
                                      contributionValue / 2.0
      }
    }
    
    edgeBetweennessScores.toList.sortBy { case (edge, score) => (-score, edge._1, edge._2) }
  }
  
  def findCommunities(workingGraph: Map[String, Set[String]], vertices: List[String]): List[List[String]] = {
    val visitedSet = MutableSet[String]()
    val componentList = ArrayBuffer[List[String]]()
    
    val remainingVertices = vertices.to[mutable.Queue]
    
    while (remainingVertices.nonEmpty) {
      val startVertex = remainingVertices.dequeue()
      
      if (!visitedSet.contains(startVertex)) {
        val componentQueue = Queue[String]()
        componentQueue.enqueue(startVertex)
        visitedSet.add(startVertex)
        val component = ArrayBuffer[String]()
        
        while (componentQueue.nonEmpty) {
          val vertex = componentQueue.dequeue()
          component += vertex
          
          for (neighbor <- workingGraph(vertex)) {
            if (!visitedSet.contains(neighbor)) {
              remainingVertices.dequeueAll(_ == neighbor)
              componentQueue.enqueue(neighbor)
              visitedSet.add(neighbor)
            }
          }
        }
        
        componentList += component.sorted.toList
      }
    }
    
    componentList.toList
  }
  
  def calculateModularity(
    communities: List[List[String]], 
    originalGraph: Map[String, Set[String]], 
    totalEdges: Int
  ): Double = {
    
    if (totalEdges == 0) return 0.0
    
    val nodeDegrees = originalGraph.map { case (node, neighbors) => (node, neighbors.size) }
    
    var modularityValue = 0.0
    
    for (component <- communities) {
      for (nodeI <- component) {
        for (nodeJ <- component) {
          val hasEdge = if (originalGraph(nodeI).contains(nodeJ)) 1.0 else 0.0
          modularityValue += hasEdge - 
                           (nodeDegrees(nodeI) * nodeDegrees(nodeJ)) / (2.0 * totalEdges)
        }
      }
    }
    
    modularityValue / (2.0 * totalEdges)
  }
  
  def main(args: Array[String]): Unit = {
    val thresholdValue = args(0).toInt
    val dataInputPath = args(1)
    val betweennessFilePath = args(2)
    val communityFilePath = args(3)
    
    val sc = new SparkContext("local[*]", "task2")
    sc.setLogLevel("ERROR")
    
    val dataLines = sc.textFile(dataInputPath)
    val headerLine = dataLines.first()
    val records = dataLines.filter(line => line != headerLine)
      .map(line => {
        val parts = line.split(",")
        (parts(0), parts(1))
      })
    
    val userToBusinesses = records.groupByKey().mapValues(_.toSet)
    val businessSetByUser = userToBusinesses.collectAsMap().toMap
    
    val allUsers = records.map(_._1).distinct().collect().toList
    
    val edgeList = ArrayBuffer[(String, String)]()
    val vertexCollection = MutableSet[String]()
    
    for (i <- allUsers.indices) {
      for (j <- i + 1 until allUsers.length) {
        val userA = allUsers(i)
        val userB = allUsers(j)
        
        val commonBusinesses = businessSetByUser(userA).intersect(businessSetByUser(userB))
        if (commonBusinesses.size >= thresholdValue) {
          edgeList += ((userA, userB))
          edgeList += ((userB, userA))
          vertexCollection.add(userA)
          vertexCollection.add(userB)
        }
      }
    }
    
    val vertices = vertexCollection.toList
    
    val adjacencyStructure = MutableMap[String, MutableSet[String]]()
    
    for ((userX, userY) <- edgeList) {
      if (!adjacencyStructure.contains(userX)) {
        adjacencyStructure(userX) = MutableSet[String]()
      }
      adjacencyStructure(userX).add(userY)
      
      if (!adjacencyStructure.contains(userY)) {
        adjacencyStructure(userY) = MutableSet[String]()
      }
      adjacencyStructure(userY).add(userX)
    }
    
    val adjacencyMap = adjacencyStructure.map { case (k, v) => (k, v.toSet) }.toMap
    
    var betweennessResults = computeEdgeBetweenness(adjacencyMap, vertices)
    
    val betweennessWriter = new java.io.PrintWriter(betweennessFilePath)
    try {
      for ((edgeTuple, score) <- betweennessResults) {
        betweennessWriter.println(s"$edgeTuple,${BigDecimal(score).setScale(5, BigDecimal.RoundingMode.HALF_UP)}")
      }
    } finally {
      betweennessWriter.close()
    }
    
    val workingGraphMutable = MutableMap[String, MutableSet[String]]()
    for ((node, neighbors) <- adjacencyMap) {
      workingGraphMutable(node) = neighbors.to[MutableSet]
    }
    
    val totalEdges = betweennessResults.length
    val nodeDegrees = adjacencyMap.map { case (node, neighbors) => (node, neighbors.size) }
    
    var bestModularityScore = Double.NegativeInfinity
    var optimalCommunities = List[List[String]]()
    
    while (betweennessResults.nonEmpty) {
      val workingGraphImmutable = workingGraphMutable.map { case (k, v) => (k, v.toSet) }.toMap
      val componentList = findCommunities(workingGraphImmutable, vertices)
      
      val modularityValue = calculateModularity(componentList, adjacencyMap, totalEdges)
      
      if (modularityValue > bestModularityScore) {
        bestModularityScore = modularityValue
        optimalCommunities = componentList
      }
      
      val maxBetweenness = betweennessResults.head._2
      for ((edge, betweennessVal) <- betweennessResults) {
        if (betweennessVal >= maxBetweenness) {
          workingGraphMutable(edge._1).remove(edge._2)
          workingGraphMutable(edge._2).remove(edge._1)
        }
      }
      
      val workingGraphForBetweenness = workingGraphMutable.map { case (k, v) => (k, v.toSet) }.toMap
      betweennessResults = computeEdgeBetweenness(workingGraphForBetweenness, vertices)
    }
    
    val sortedCommunities = optimalCommunities.sortBy(c => (c.size, c.head))
    s
    val communityWriter = new java.io.PrintWriter(communityFilePath)
    try {
      for (community <- sortedCommunities) {
        communityWriter.println(community.map(u => s"'$u'").mkString(", "))
      }
    } finally {
      communityWriter.close()
    }
    
    sc.stop()
  }
}